json: Improve tolerance to uncovered C members during JSON encoding

lucasdietrich opened the pull request:
This enhancement fixes encoding of JSON objects for which descriptors do not covers all members of the C structure represented.

Fix #50976

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

carlescufi commented:
@mrfuchs could you please take a look? Thanks!

MaureenHelm commented:
@M1cha can you take a look?

M1cha commented on include/zephyr/data/json.h:
```diff
@@ -137,6 +138,9 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 				 __alignof__(type) == 4 ? 2 : 3)
 
+/* Helper to get the size of a member within a struct. */
+#define Z_MEMBER_SIZE(type, member) sizeof(((type *)0)->member)
```
Given how often I needed this in other zephyr projects I'd say it's finally time to add this to `include/zephyr/sys/util.h` with the name `SIZEOF_FIELD` (that's what the linux kernel calls this, just lowercase).

M1cha commented on include/zephyr/data/json.h:
```diff
@@ -109,7 +109,8 @@ struct json_obj_descr {
 		struct {
 			const struct json_obj_descr *sub_descr;
-			size_t sub_descr_len;
+			uint16_t sub_descr_len;
+			uint16_t sub_struct_size;
```
what's the reason for the type change and why did you choose `uint16_t`?

lucasdietrich replied:
Reverted the type for `sub_descr_len` to `size_t`.

I do agree, I moved the field `sub_struct_size` up and renamed it to `struct_size`. It just required a little more work to have all descriptors defined correctly: I needed to explicitly define the `struct_size` member to an arbitrary value (e.g. 0).

lucasdietrich commented on tests/lib/json/src/main.c:
```diff
@@ -255,10 +264,11 @@ ZTEST(lib_json_test, test_json_limits)
 ZTEST(lib_json_test, test_json_decoding_array_array)
 {
 	int ret;
-	struct obj_array_array obj_array_array_ts;
```
Indeed, reverted that

carlescufi commented:
@mrfuchs please take a look

mrfuchs commented on lib/os/json.c:
```diff
@@ -549,6 +540,8 @@ static int arr_parse(struct json_obj *obj,
 
 	if (val) {
 		elements = (size_t *)((char *)val + elem_descr->offset);
+	} else {
+		field = (char *)field + elem_descr->offset;
```
When the `field` pointer is updated and skipped over a (unused?) field, shouldn't the `last_elem` variable also be updated?

lucasdietrich replied:
In this new design `get_elem_size()` returns the actual size of the underlying object, so `last_elem` correctly reflects the address of the last element.

However I'm still not convinced the way `field` is adjusted when parsing the nested array (when parsing an array of array).

mrfuchs replied:
> However I'm still not convinced the way `field` is adjusted when parsing the nested array (when parsing an array of array).

Indeed. I hope that'll be fixed soon by #50816.

mrfuchs commented on include/zephyr/sys/util.h:
```diff
@@ -210,6 +210,26 @@ extern "C" {
+ * @param member the name of the field in the structure
+ * @return the size of the field in bytes
+ */
+#define SIZEOF_FIELD(type, member) sizeof(((type *)0)->member)
```
I have no strong opinion on this naming choice, but looking at the other macro names, shouldn't it be called `FIELD_SIZE` (cf. `ARRAY_SIZE`)?

mrfuchs replied:
Ah, sorry, I just saw that this names comes from @M1cha's advice with reference to the Linux kernel.

lucasdietrich commented:
Did several modifications, but I still need to do additional testing.

pdgendt commented:
The `align_shift` member becomes obsolete if the struct size is to be used.

pdgendt commented on include/zephyr/data/json.h:
```diff
@@ -106,6 +106,9 @@ struct json_obj_descr {
 	uint32_t offset : 16;
 
+	/* Size of the structure being described by the "object" descriptor */
+	size_t struct_size;
```
Remove the `align_shift` member. This frees up 2 bits, these can be used to `field_name_len` and `type` to get an extra bit?

pdgendt commented on include/zephyr/sys/util.h:
```diff
@@ -210,6 +210,26 @@ extern "C" {
+ * @param member the name of the field in the structure
+ * @return the size of the field in bytes
+ */
+#define SIZEOF_FIELD(type, member) sizeof(((type *)0)->member)
```
This should be added in a separate commit.

pdgendt commented on tests/lib/json/src/main.c:
```diff
@@ -11,6 +11,7 @@
 
 struct test_nested {
 	int nested_int;
+	uint32_t _unused_member;
```
Create a separate commit for the changes in tests.

pdgendt commented:
@lucasdietrich care to update the PR for the latest comments, thanks

The tests in my PR #59485 succeed using this one.

lucasdietrich commented:
Thank you for your review, I will make the changes this evening.

lucasdietrich commented:
I believe I have addressed all your remarks @mrfuchs @pdgendt.

lucasdietrich committed:
sys: Introduce SIZEOF_FIELD helper macro in util.h

This macro is used to get the size of a specific field in a structure type.
It will return the size of the field in bytes.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

lucasdietrich committed:
test: Add unused members to test structs in JSON test suite

This commit updates various structs in the JSON test suite to include
unused members. These additions aim to test the improved tolerance of
JSON encoding for C struct members that may not be fully described by
JSON descriptors.

As a consequence of the addition of unused fields,
prefer designated initializers instead of postionnal ones.

Add a note about a test which constantly fails with a unused field.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

pdgendt commented:
@lucasdietrich #50816 has been merged, please rebase and resolve conflicts, thanks!

lucasdietrich commented:
Ok, i will rebase this evening.

lucasdietrich committed:
json: Introduce struct_size descriptor member

Enable the computation of a struct member's size during compile-time to
enhance accuracy, removing the need for speculative size determinations
during runtime.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

lucasdietrich commented:
Unfortunately I didn't have time to work on it yet, moreover the merge seems to be tricky.
I will need few more days.

pdgendt commented:
@lucasdietrich any update on this?

lucasdietrich commented:
After more investigation, I managed to get almost everything working except for the test "test_json_decoding_array_array", which has become a blocker for me.

I've somewhat lost track of the issue, but I can redirect my attention to it by tomorrow evening.

However, I might require some help to resolve it completely.

lucasdietrich commented:
I managed to get almost everything working; we don't "guess" anymore the size of a structure but we compute it at build time and we use this to determine offsets.

The most tricky part was (and still is) the handling of nested arrays (array of arrays), which requires dedicated handling I'm not able to fully explain. To be honest I tried to have the unit tests working rather than designing an accurate solution for nested arrays.  This is where I would greatly benefit from external insight. Consequently this has left the following key issue unresolved : 

Yet, the member of the structure describing the nested array should remain at the start of the structure. Using the "test_json_encoding_array_array" test as example: I'm not yet able to calculate the correct offset of "objects" within the "struct array". Consequently the test fails (crashes) if we uncomment the line. Actually I don't fully understand when the adjustment should be made.

I've successfully tested these changes with a personnel project.

lucasdietrich commented:
I've just rebased and noticed that the new tests introduced in #60046 are failing with the current changes(2 dim array encoding). I'll need more time to address this issue.

lucasdietrich committed:
json: Improve tolerance to uncovered C members during JSON encoding

Add special handling for nested arrays

This enhancement fix encoding of JSON objects for which descriptors do
not covers all members of the C structure represented.

Fixes #50976

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>
